From 5ee2e04db4ec48c049518f9bb5a126e66bd3d391 Mon Sep 17 00:00:00 2001
From: Valtteri Vuorikoski <vuori@notcom.org>
Date: Tue, 21 Apr 2026 22:26:18 +0300
Subject: [PATCH] commands/pgp: fix new-format packet length parsing.

Parsing of OpenPGP "new format" packet lengths is incorrect for the 2-octet
format. The problem was originally reported in the Savannah ticket
https://savannah.gnu.org/bugs/?68020. This applies the proposed fix and adds
some explanatory comments because the length encoding is quite complicated.

Sequoia-PGP produces the new format, while GnuPG produces the old format.

Signed-off-by: Valtteri Vuorikoski <vuori@notcom.org>
Reported-by: shoesoft <>
Reviewed-by: Andrew Hamilton <adhamilt@gmail.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/107>

Index: 5ee2e04d/grub-core/commands/pgp.c
===================================================================
--- 5ee2e04d.orig/grub-core/commands/pgp.c
+++ 5ee2e04d/grub-core/commands/pgp.c
@@ -79,7 +79,7 @@ read_packet_header (grub_file_t sig, gru
 
   if (!(type & 0x80))
     return grub_error (GRUB_ERR_BAD_SIGNATURE, N_("bad signature"));
-  if (type & 0x40)
+  if (type & 0x40) /* Bit 6 set: new format packet length */
     {
       *out_type = (type & 0x3f);
       if (grub_file_read (sig, &l, sizeof (l)) != 1)
@@ -91,10 +91,11 @@ read_packet_header (grub_file_t sig, gru
 	}
       if (l < 224)
 	{
+          /* RFC4880 2-octet length: ((b1 - 192) << 8) + (b2 + 192) */
 	  *len = (l - 192) << GRUB_CHAR_BIT;
 	  if (grub_file_read (sig, &l, sizeof (l)) != 1)
 	    return grub_error (GRUB_ERR_BAD_SIGNATURE, N_("bad signature"));
-	  *len |= l;
+	  *len = (*len | l) + 192;
 	  return 0;
 	}
       if (l == 255)
@@ -106,6 +107,8 @@ read_packet_header (grub_file_t sig, gru
 	}
       return grub_error (GRUB_ERR_BAD_SIGNATURE, N_("bad signature"));
     }
+
+  /* Bit 6 unset: old format packet length */
   *out_type = ((type >> 2) & 0xf);
   switch (type & 0x3)
     {
